src/python/pants/engine/internals/defaults.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Iterable, Mapping, Tuple, Union
+
+from pants.engine.addresses import Address
+from pants.engine.target import ImmutableValue, InvalidFieldException, RegisteredTargetTypes
+from pants.engine.unions import UnionMembership
+from pants.util.frozendict import FrozenDict
+
+SetDefaultsValueT = Mapping[str, Any]
+SetDefaultsKeyT = Union[str, Tuple[str, ...]]
+SetDefaultsT = Mapping[SetDefaultsKeyT, SetDefaultsValueT]
+
+
+@dataclass
+class BuildFileDefaultsProvider:
+    registered_target_types: RegisteredTargetTypes
+    union_membership: UnionMembership
+
+    def get_parser_defaults(
+        self, path: str, defaults: BuildFileDefaults
+    ) -> BuildFileDefaultsParserState:
+        return BuildFileDefaultsParserState(
+            address=Address(path, generated_name="__defaults__"),
+            defaults=dict(defaults),
+            provider=self,
+        )
+
+
+class BuildFileDefaults(FrozenDict[str, FrozenDict[str, ImmutableValue]]):
+    """Map target types to default field values."""
+
+
+@dataclass
+class BuildFileDefaultsParserState:
```

Comment:
A class doc or (module doc?) explaining how defaults are applied would be helpful.

Reply:
Oh, whoops... I think that I actually meant implementation details rather than usage details.

But the docs that you've written are actually the perfect reminder that we should be writing docs for new features around the time when we land them. So maybe you can move the very good docstring that you added above to a section in https://github.com/pantsbuild/pants/blob/main/docs/markdown/Using%20Pants/concepts/targets.md ?